os/generate-self-signed-certificates.md

```diff
@@ -0,0 +1,301 @@
+```
+
+### Certificate types which are used inside CoreOS
+
+* **client certificate** is used to authenticate client by server. For example `etcdctl`, `etcd proxy`, `fleetctl` or `docker` clients.
+* **server certificate** is used by server and verified by client for server identity. For example `docker` server or `kube-apiserver`.
+* **client-server certificate** is used by etcd cluster members as they communicate with each other in both ways.
+
+### Configure CA options
+
+Now we can configure signing options inside `ca-config.json` config file. Default options conatin following preconfigured fields:
+
+* profiles: **www** with `server auth` (TLS Web Server Authentication) X509 V3 extension and **client** with `client auth` (TLS Web Client Authentication) X509 V3 extension.
+* expiry: with `8760h` default value (or 365 days)
+
+For compliance let's rename **www** profile into **server**, create additional **client-server** profile with both `server auth` and `client auth` extensions, and set expiry to 43800h (5 years):
+
+```json
+{
+    "signing": {
+        "default": {
+            "expiry": "43800h"
+        },
+        "profiles": {
+            "server": {
+                "usages": [
+                    "signing",
+                    "key encipherment",
+                    "server auth"
+                ]
+            },
+            "client": {
+                "usages": [
+                    "signing",
+                    "key encipherment",
+                    "client auth"
+                ]
+            },
+            "client-server": {
+                "usages": [
+                    "signing",
+                    "key encipherment",
+                    "server auth",
+                    "client auth"
+                ]
+            }
+        }
+    }
+}
+```
+
+You can also modify `ca-csr.json` Certificate Signing Request (CSR):
+
+```json
+{
+    "CN": "My own CA",
+    "key": {
+        "algo": "rsa",
+        "size": 2048
+    },
+    "names": [
+        {
+            "C": "US",
+            "L": "CA",
+            "O": "My Company Name",
+            "ST": "San Francisco",
+            "OU": "Org Unit 1",
+            "OU": "Org Unit 2"
+        }
+    ]
+}
+```
+
+And generate CA with defined options:
+
+```sh
+cfssl gencert -initca ca-csr.json | cfssljson -bare ca -
+```
+
+Or generate CA with **default** options:
+
+```sh
+cfssl print-defaults csr | cfssl gencert -initca - | cfssljson -bare ca -
+``` 
+
+You'll get following files:
+
+```
+ca-key.pem
+ca.csr
+ca.pem
+```
+
+* Please keep `ca-key.pem` file in safe, i.e. on your local host. This key allows to create any kind of certificates within your CA.
+* **\*.csr** files are not used and you can delete them.
+
+### Generate server certificate
+
+```
+cfssl print-defaults csr > server.json
+```
+
+Most important values for server certificate are **Common Name (CN)** and **hosts**. We have to substitute them, for example:
```

Comment:
may-be-related: I know some users want to have wild-card server certificate so they could use same server ceritificate in all servers. unsure about whether we need to support/guide that.

@gtank Thoughts?